enhancements/insights/pulling-data-from-ocm.md

```diff
@@ -0,0 +1,133 @@
+last-updated: 2021-03-09
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Insights Operator pulling and exposing data from the OCM API
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement will enable the Insights Operator to pull the data (SCA certs)
+from the OCM (OpenShift Cluster Manager) API. The data will be exposed by the Insights Operator
+in the OpenShift API to allow users to use them when consuming and building container images
+on the platform.
+
+## Motivation
+
+Users could consume RHEL content and container images using the RHEL subscription in the OpenShift 3.x.
+In the OpenShift 4, this is no longer possible because the Red Hat Enterprise Linux Core OS (RHCOS) does not
+provide any attached subscription. This enhancement is to provide users the Simple Content Access (SCA) certs
+from Red Hat Subscription Manager (RHSM).
+
+### Goals
+
+- Extend the Insights Operator config with an OCM API URL to be able to query the data
+- Periodically pull the data from the OCM API and expose it in the OpenShift API
+
+### Non-Goals
+
+- Insights Operator providing any transformation or post-processing of the data pulled
+  from the OCM API
+
+## Proposal
+
+### User Stories
+
+#### Consume SCA certs exposed in the API
+
+As an OpenShift user
+I want to consume SCA certs to be able to consume RHEL content and to build
+corresponding container images.
+
+### Risks and Mitigations
+
+#### OCM API is down
+
+Risk: OCM API is down or doesn't provide up to date data.
+
+Risk: Insights Operator is unable to expose/update the data in the OpenShift API
+
+Mitigation: Introduce a new state in the Insights Operator (e.g "SCADataDegraded") and
+create a new alert based on this new state.
+
+## Design Details
+
+### Authorization
+
+The Insights Operator is able to pull the data from the OCM API using the existing `cloud.openshift.com` token
+available in the `pull-secret` (in the `openshift-config-managed` namespace).
+
+The Insights Operator must provide a cluster ID as an identifier of the cluster.
+
+### Data in API
+
+The SCA certificate is available via the `etc-pki-entitlement` secret in the `openshift-config-managed` namespace.
+
+### Update period
+- Insights Operator query the OCM API every 8 hours and downloads the full data provided
+
+### Test Plan
+
+- `insights-operator-e2e-tests` suite can verify the SCA cert data
+  is available
+- Basic test of the validity of the SCA certs. Mount the `etc-pki-entitlement` secret and run e.g `yum install` in the container
+
+### Graduation Criteria
+
+This feature is planned as a technical preview in OCP 4.9 and is planned to go GA in 4.10.
```

Comment:
does this mean the feature will be disabled by default and customers can opt-in using some feature gate?

Reply:
Sorry maybe this is too vaguely defined, but it will be the opposite. It's enabled by default and users/customers can disable it.

Reply:
Note that a lot of the mechanisms we need to get this all working together (projected resource CSI driver, builds with CSI inline volumes) are going to be true tech preview features - no upgrade, must opt in. I recommend the same - use a feature gate to enable this capability, add it to the `TechPreviewNoUpgrade` feature set.

@deads2k @s-urbaniak I don't think we can get the CSI bits to GA until we have better SCC/admission control around CSI inline volumes.

Reply:
OK so I think I would need some help here please. I don't know how a feature gate works or what it exactly means. Also I don't know about the feature set.

Reply:
You make a feature gate in 4.9 and say "this is tech preview, if you want to opt-in, you have to enable this"... then in 4.10 you just make the feature gate on by default (assuming everything will work fine in clusters that have this enabled).

Reply:
Feature gates are free-form strings that can be added to the `FeatureGate` API object (a custom resource specific to OpenShift). We have facilities to group feature gates into feature sets so cluster admins don't need to know all of them to enable logical groups features.

See https://github.com/openshift/api/blob/master/config/v1/types_feature.go#L113-L121

Reply:
OK thanks. I opened https://github.com/openshift/api/pull/983

Reply:
this discussion appears to be resolved, i am marking it so.